Implementaion  for stationary sources of pollution

pronapro opened the pull request:
## Description

[Implementation for API endpoints for stationary sources of pollution ]

## Changes Made

- [x] Made API endpoints for Data insertion and retrieval.
- [x] Data insertion endpoints take in the image, duration of interest and radius of area of interest and return the geometry, confidence score and location and environment profiles identified.
- [x] Retrieving endpoints return data saved in the database by parameters such as confidence score.

## Testing

- [x] Tested locally
- [ ] Tested against staging environment
- [ ] Relevant tests passed: [List test names]

## Affected Services

- [ ] Which services were modified:
  - [ ] Spatial microservice
  - [ ] 
## Additional Setup Notes
### Pre-requisites
- **Google Cloud Credentials**: Ensure you have your `credentials.json` file from Google Cloud. Place this file in the root directory of your project.
- **Environment Variables**: Create a `.env` file in the root directory with necessary environment variables (e.g., API keys, database URLs).

### Running the Docker Container
1. Build the Docker image:
`docker build -t your-image-name .`
2. Run the Docker container:
`docker run -p 5000:5000 your-image-name`


### Testing with Postman
- Use Postman to test the API endpoints.
### API Endpoints
Ensure you have Postman installed and set up to test the following API endpoints. Replace `http://localhost:4000` with your actual server address if different.

#### Upload Image for Prediction
- **POST** `/upload-image` Use this endpoint to upload an image file for prediction. Ensure the request body type is set to form-data and includes a file with the key 'file'.

#### Get Data by Confidence
- **GET** `/get-data-by-confidence?min_confidence=0.7` This endpoint retrieves data filtered by confidence score.
#### Save All Location Data
- **GET** `/save-all-location` This endpoint triggers the addition of all location profiles to the database.
#### Save All Environment Data
- **GET** `/save-all-environment_data`  This endpoint triggers the addition of all environmental profiles to the database.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced new API endpoints for uploading images and retrieving pollutant data based on confidence levels.
  - Added functionality for processing and predicting pollutant presence using uploaded images.
  - Enhanced configuration options for MongoDB connection settings.
  - Comprehensive methods for managing pollutant data through image uploads and MongoDB interactions.
  - Added support for loading TensorFlow Lite models from Google Cloud Storage.

- **Bug Fixes**
  - Improved error handling for data insertion and retrieval processes.

- **Documentation**
  - Updated requirements to reflect new dependencies and specific versioning for better stability.
  - Enhanced file exclusion criteria in the repository.

These updates enhance the application's ability to process geospatial data and improve user interaction with pollutant data.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Baalmart committed:
Merge pull request #3907 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3909 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3911 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3913 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3915 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3917 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3920 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3924 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3926 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3929 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3930 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3932 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3933 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3935 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3939 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3940 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3942 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3944 from airqo-platform/staging

move to production

pronapro committed:
Implentation for stationary sources of pollution

Co-Authored-By: Noble Mutabazi <60974514+Mnoble-19@users.noreply.github.com>

Baalmart committed:
Merge pull request #3947 from airqo-platform/staging

move to production

Baalmart reviewed (changes requested):
Thanks. @pronapro for this. The future is indeed bright.

- I can see that the deploy preview for the PR is failing. Please investigate this.
- Can you please reach out (individually) to @wabinyai and @Mnoble-19 to thoroughly review this PR?

Thanks again.

pronapro committed:
Dependence issues

Baalmart committed:
Merge pull request #3949 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3951 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3953 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3955 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3957 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3959 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3961 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3963 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3965 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3967 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3969 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3971 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3973 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3975 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3977 from airqo-platform/staging

[work in progress] move to production

Baalmart committed:
Merge pull request #3980 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3982 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3984 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3986 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3989 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3991 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3992 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3994 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #3997 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4000 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4002 from airqo-platform/staging

Baalmart committed:
Merge pull request #4004 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4008 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4010 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4012 from airqo-platform/staging

Baalmart committed:
Merge pull request #4014 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4016 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4019 from airqo-platform/staging

Move to production

Baalmart committed:
Merge pull request #4021 from airqo-platform/staging

Move to production

Baalmart committed:
Merge pull request #4023 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4028 from airqo-platform/staging

Move to production

Baalmart committed:
Merge pull request #4030 from airqo-platform/staging

Move to production

Baalmart committed:
Merge pull request #4032 from airqo-platform/staging

[move to production

Baalmart committed:
Merge pull request #4036 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4039 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4041 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4043 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4046 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4048 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4051 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4053 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4055 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4057 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4059 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4061 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4063 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4065 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4067 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4070 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4071 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4073 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4075 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4077 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4079 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4081 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4084 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4086 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4088 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4092 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4093 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4095 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4097 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4099 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4101 from airqo-platform/staging

Move to production

Baalmart committed:
Merge pull request #4103 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4105 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4107 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4109 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4112 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4114 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4116 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4119 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4121 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4123 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4125 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4127 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4129 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4131 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4133 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4135 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4138 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4141 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4143 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4144 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4145 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4147 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4150 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4154 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4156 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4160 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4164 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4166 from airqo-platform/staging

 move to production

Baalmart committed:
Merge pull request #4168 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4171 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4173 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4176 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4178 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4180 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4182 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4183 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4185 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4188 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4189 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4191 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4193 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4195 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4199 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4200 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4202 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4205 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4212 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4215 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4218 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4220 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4222 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4226 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4228 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4230 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4232 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4234 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4236 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4238 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4241 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4243 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4245 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4248 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4250 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4252 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4256 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4259 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4261 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4263 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4265 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4267 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4269 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4271 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4273 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4276 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4278 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4279 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4281 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4284 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4286 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4293 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4295 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4297 from airqo-platform/staging

 move to production

Baalmart committed:
Merge pull request #4299 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4301 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4303 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4305 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4308 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4313 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4315 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4317 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4319 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4321 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4323 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4325 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4328 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4329 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4331 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4333 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4335 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4337 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4339 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4341 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4342 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4344 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4346 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4348 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4350 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4352 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4354 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4356 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4358 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4360 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4362 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4364 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4367 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4368 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4371 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4373 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4376 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4378 from airqo-platform/staging

move to production

pronapro committed:
Refactored code with optimized model

Baalmart committed:
Merge pull request #4380 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4385 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4386 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4388 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4390 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4393 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4395 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4397 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4399 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4401 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4404 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4406 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4409 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4411 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4413 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4415 from airqo-platform/staging

move to production

Baalmart committed:
Merge pull request #4417 from airqo-platform/staging

move to production

pronapro committed:
Uncommented code

pronapro committed:
Merge branch 'master' of https://github.com/airqo-platform/AirQo-api into pollutantimplementation

pronapro committed:
updated env variables

pronapro committed:
Error handling

pronapro committed:
Refactored code

Removed the profile information codes and inline debugging code(print statements).

pronapro committed:
Code refactor PT2

pronapro committed:
Profile file

pronapro committed:
Fetching cordinates and ids functionality

pronapro committed:
Location profile functions and database

pronapro committed:
Location profile error

pronapro committed:
Environment functions added

pronapro committed:
removed logging and exposed exceptions

wabinyai committed:
tre

wabinyai committed:
po

wabinyai committed:
sate

wabinyai commented on src/spatial/.dockerignore:
```diff
@@ -17,7 +17,7 @@ production.env
 
 # datafiles
-google_application_credentials.json
+#google_application_credentials.json
```
I do not think this needs to be removed.  We do not want to accidentally add these to the PR

wabinyai commented on src/spatial/requirements.txt:
```diff
@@ -1,28 +1,40 @@
 geojson~=3.1.0
 libpysal~=4.12.1
 requests~=2.32.3
-python-dotenv~=1.0.1
```
**python-dotenv** Is used by  other models in this microservice keep it

wabinyai commented on src/spatial/requirements.txt:
```diff
@@ -1,28 +1,40 @@
 google-cloud-bigquery[pandas] 
-gunicorn
 flask-cors~=5.0.0
-earthengine-api~=1.1.1
```
**earthengine-api~=1.1.1** do not remove this. It is also being used in the microservice, especially to pull satellite data.

wabinyai commented on src/spatial/.dockerignore:
```diff
@@ -2,7 +2,7 @@
 
 # Environments 
-.env
+#.env
```
Do not comment out **.env** is being used, we do not need to accidentally display out secrets

wabinyai reviewed (commented):
Sort out these comments

pronapro committed:
change of database

pronapro committed:
refactored docker ignore file

pronapro committed:
Merge branch 'pollutantimplementation' of https://github.com/airqo-platform/AirQo-api into pollutantimplementation

pronapro committed:
updated requirements file

pronapro committed:
Google building data and land cover added

pronapro committed:
Merge branch 'staging' into pollutantimplementation

sserurich reviewed (commented):
some feedback from code scanning related to improving security (when printing exception messages) , can address them now or we can initiate other PRs to improve security of the codebase
![image](https://github.com/user-attachments/assets/88f65729-9aa8-4fd4-b8fc-b9003a83deac)

sserurich reviewed (approved):
The functionality works as expected. Tested locally, left a comment but doesn't stop approving.

pronapro committed:
Merge remote-tracking branch 'origin' into pollutantimplementation

wabinyai reviewed (approved):
Looks ok, the comments I left were addressed
